Fixed remaining platform/apisupport tests on JDK11+

matthiasblaesing opened the pull request:
... the final chapter?

matthiasblaesing commented:
The final full CI run ran mostly clean, just "Web Modules" needed a manual restart. As that is not related to these changes, I would put it into the "normal flakiness" category.

mbien commented on platform/openide.nodes/src/org/openide/nodes/Children.java:
```diff
@@ -1942,15 +1942,9 @@ private Mutex callPMMutexMethod() {
-            } catch (InvocationTargetException e) {
-                return FALLBACK;
-            } catch (NoSuchMethodException e) {
+            } catch (ExceptionInInitializerError | ClassNotFoundException
+                    | IllegalAccessException | IllegalArgumentException
+                    | InvocationTargetException | NoSuchMethodException e) {
```
nitpick: `ReflectiveOperationException | ExceptionInInitializerError | IllegalArgumentException`

mbien commented on platform/openide.util.ui/test/unit/src/org/openide/xml/XMLUtilTest2.java:
```diff
@@ -49,9 +49,9 @@
 import org.xml.sax.helpers.DefaultHandler;
 
-public class XMLUtilTest extends NbTestCase {
+public class XMLUtilTest2 extends NbTestCase {
```
adding the `2` does disable the test?

matthiasblaesing replied:
The 2 was intended to give the class a different name. I overlooked, that the class is not executed anymore. Will be fixed.

mbien commented on platform/openide.util.ui/test/unit/src/org/openide/xml/XMLUtilTest2.java:
```diff
@@ -364,7 +364,11 @@ public void testSignificantWhitespace() throws Exception {
         String data =
                 "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" +
                 "<r>\n" +
-                "    <p>This is <em>not</em> a test!</p>\n" +
```
i remember looking at this before and I couldn't figure out if the test was supposed to assert that NB isn't re-formatting the code or if is about something else

matthiasblaesing replied:
F*** we might indeed see a Xalan bug here. Not sure what to make of this.

mbien replied:
i have a few attempts on those tests in my stash. I remember trying to influence how the whitespace is handled. Let me finish the maven PR I am working on first, then I take a look at the stash.

matthiasblaesing replied:
Would be interesting. At this point I'm inclined to say, that the default ident feature of xalan does not preserve significant white space and the solution might be to not let the xlst processor do the indention, but do it ourselves.

matthiasblaesing replied:
I found what triggered my change. From the documentation of `XMLUtil#write`:

> The precise output format is not guaranteed but this method will attempt to indent it sensibly.

My fix is still broken, but ....

mbien replied:
so i looked at this probably 2 years ago, but I remembered a few things after looking at the stash:
 - JDK 9-13 handles whitespace in CDATA differently, `testCDATA()` test fails only in that range -> my suggestion: lets forget about JDK 11 and go right to 17+
 - the `testSignificantWhitespace()` fails from 9+
 - `XMLUtil.parse` / `XMLUtil.write` **is always** reformatting the document. The difference is that on 8, a tag+text sequence does not cause a newline, but on 17+ it does. I suppose this is what "Significant Whitespace" means? Never heard of it before but i got a few hits - will take another look some time later.

mbien replied:
https://bugs.openjdk.org/browse/JDK-8223291 this is likely the CDATA issue, it matches the 9-13 JDK range

mbien replied:
I am to 90% sure that there is no real solution to this. Since the method would have to detect mixed content and switch from xml to html mode to be able to make the test happy. The doc on top of the method doesn't really specify how it formats anything.

Since NB is working fine on JDK 9+ this implementation detail of mixed content formatting doesn't seem to be a problem.

Lets turn `testSignificantWhitespace()` off, and get this in :)

mbien commented:
nitpick: can we rename the PR to "Fixed remaining platform tests on JDK11+" or similar? Makes it easier to find it again among the many other PRs which did similar things in other clusters.

matthiasblaesing commented:
I failed to determine a sane way to define what is "significant whitespace" in the absence of a DTD/XSD. The definition I found was "without a DTD/XSD" any whitespace is significant. I assume "any" in this case means "outside" the tags themselves. As that runs contrary to the method documentation, I agree, that the assumption about the methods behavior was just wrong. So I removed the test.

I applied the exception cleanup and fixed the naming of the XmlUtil2Test, so that it is executed aain.

mbien reviewed (approved):
read the commit booklets and the changes looked good to me, great work!

only beauty mistake I noticed is that
if you compare https://github.com/apache/netbeans/pull/7502/commits/88a5a8e704f67258ff027bbaa8d2608074ea5574 and https://github.com/apache/netbeans/pull/7502/commits/6440e808f203cdc98cbc0418db433ee69cccc91b you see that one duplicates `platform/openide.text` in `main.yml` the other overwrites it. This usually indicates that something was squashed to the wrong commit. But the final result of all commits looks good.

matthiasblaesing commented:
@mbien thanks for the review. Indeed I rebased multiple times and it seems something got messed up. I'll leave it as is, as even in worst case a unittestset is executed twice on GH.
